src/plugins/part.c

```diff
@@ -1000,107 +1085,370 @@ static PedPartition* add_part_to_disk (PedDevice *dev, PedDisk *disk, BDPartType
+        if (status != 0) {
+            g_set_error (error, BD_PART_ERROR, BD_PART_ERROR_FAIL,
+                         "Failed to get new partition number");
+            fdisk_unref_table (table);
+            fdisk_unref_partition (npa);
+            close_context (cxt);
             bd_utils_report_finished (progress_id, (*error)->message);
             return NULL;
         }
+    }
 
-        if (ped_part->type == PED_PARTITION_EXTENDED) {
-            /* can at minimal start where the first logical partition can start - the start of the extended partition + 1 MiB aligned up */
-            if (start < ((ped_part->geom.start * dev->sector_size) + 1 MiB + dev->sector_size - 1))
-                start = (ped_part->geom.start * dev->sector_size) + 1 MiB + dev->sector_size - 1;
-        } else {
-            /* can at minimal start where the next logical partition can start - the end of the previous partition + 1 MiB aligned up */
-            if (start < ((ped_part->geom.end * dev->sector_size) + 1 MiB + dev->sector_size - 1))
-                start = (ped_part->geom.end * dev->sector_size) + 1 MiB + dev->sector_size - 1;
-        }
+    status = fdisk_partition_set_partno (npa, partno);
+    if (status != 0) {
+        g_set_error (error, BD_PART_ERROR, BD_PART_ERROR_FAIL,
+                     "Failed to set new partition number");
+        fdisk_unref_table (table);
+        fdisk_unref_partition (npa);
+        close_context (cxt);
+        bd_utils_report_finished (progress_id, (*error)->message);
+        return NULL;
     }
 
-    ped_part = add_part_to_disk (dev, ped_disk, type, start, size, align, error);
-    if (!ped_part) {
-        /* error is already populated */
-        ped_disk_destroy (ped_disk);
-        ped_device_destroy (dev);
+    status = fdisk_add_partition (cxt, npa, NULL);
+    if (status != 0) {
+        g_set_error (error, BD_PART_ERROR, BD_PART_ERROR_FAIL,
+                     "Failed to add new partition to the table: %s", strerror_l (-status, c_locale));
+        fdisk_unref_table (table);
+        fdisk_unref_partition (npa);
+        close_context (cxt);
+        bd_utils_report_finished (progress_id, (*error)->message);
+        return NULL;
+    }
+
+    /* for new extended partition we need to force reread whole partition table with
+       libfdisk < 2.36.1 */
+    if (!write_label (cxt, table, disk, new_extended && fdisk_version < 2361, error)) {
         bd_utils_report_finished (progress_id, (*error)->message);
+        fdisk_unref_table (table);
+        fdisk_unref_partition (npa);
+        close_context (cxt);
         return NULL;
     }
 
-    succ = disk_commit (ped_disk, disk, error);
-    if (succ)
-        ret = get_part_spec (dev, ped_disk, ped_part, error);
+    if (fdisk_partition_has_partno (npa)) {
+        if (isdigit (disk[strlen(disk) - 1]))
+            ppath = g_strdup_printf ("%sp%zu", disk, fdisk_partition_get_partno (npa) + 1);
```

Comment:
Okay, I think this deserves more testing on _exotic_ drives: (e)MMC, NVMe, NVMeOF as the device node naming is rather non-traditional there.

Reply:
I think this should be OK for all devices. Yes, there are some weird names for NVMes and similar devices, but this should not affect partition names. The device name + number (or + "p" + number for device names ending with a digit) system should work everywhere -- it works this way even for "non-disk" devices like loop or md.